File: app/services/chatwoot_client.py
"""Chatwoot API client — mirrors all HTTP calls from the n8n workflow."""
import httpx
from app.config import get_settings

settings = get_settings()


def _headers() -> dict:
    return {
        "api_access_token": settings.CHATWOOT_API_TOKEN,
        "Content-Type": "application/json",
    }


def _base(account_id: int) -> str:
    return f"{settings.CHATWOOT_URL}/api/v1/accounts/{account_id}"


async def mark_as_read(account_id: int, conversation_id: int) -> None:
    """POST update_last_seen — marks messages as read."""
    url = f"{_base(account_id)}/conversations/{conversation_id}/update_last_seen"
    async with httpx.AsyncClient() as client:
        await client.post(url, headers=_headers())


async def send_message(account_id: int, conversation_id: int, content: str) -> dict:
    """POST a text message to a conversation."""
    url = f"{_base(account_id)}/conversations/{conversation_id}/messages"
    async with httpx.AsyncClient() as client:
        r = await client.post(url, headers=_headers(), json={"content": content})
        r.raise_for_status()
        return r.json()


async def send_reaction(
    account_id: int,
    conversation_id: int,
    message_id: int,
    emoji: str,
) -> dict:
    """POST a reaction to a specific message."""
    url = f"{_base(account_id)}/conversations/{conversation_id}/messages"
    body = {
        "content": emoji,
        "content_attributes": {
            "in_reply_to": message_id,
            "is_reaction": True,
        },
    }
    async with httpx.AsyncClient() as client:
        r = await client.post(url, headers=_headers(), json=body)
        r.raise_for_status()
        return r.json()


async def update_labels(account_id: int, conversation_id: int, labels: list[str]) -> None:
    """POST/replace conversation labels."""
    url = f"{_base(account_id)}/conversations/{conversation_id}/labels"
    async with httpx.AsyncClient() as client:
        await client.post(url, headers=_headers(), json={"labels": labels})


async def get_contact(account_id: int, contact_id: int) -> dict:
    """GET contact details (includes custom_attributes)."""
    url = f"{_base(account_id)}/contacts/{contact_id}"
    async with httpx.AsyncClient() as client:
        r = await client.get(url, headers=_headers())
        r.raise_for_status()
        return r.json()


async def update_contact_attributes(
    account_id: int,
    contact_id: int,
    custom_attributes: dict,
) -> dict:
    """PATCH contact custom attributes."""
    url = f"{_base(account_id)}/contacts/{contact_id}"
    async with httpx.AsyncClient() as client:
        r = await client.patch(
            url,
            headers=_headers(),
            json={"custom_attributes": custom_attributes},
        )
        r.raise_for_status()
        return r.json()


async def destroy_contact_attributes(
    account_id: int,
    contact_id: int,
    attributes: list[str],
) -> None:
    """POST destroy_custom_attributes."""
    url = f"{_base(account_id)}/contacts/{contact_id}/destroy_custom_attributes"
    async with httpx.AsyncClient() as client:
        await client.post(
            url, headers=_headers(), json={"custom_attributes": attributes}
        )


async def download_attachment(data_url: str) -> bytes:
    """Download a binary attachment (audio, file) from Chatwoot storage."""
    async with httpx.AsyncClient() as client:
        r = await client.get(data_url, headers=_headers(), follow_redirects=True)
        r.raise_for_status()
        return r.content


async def search_contact_by_phone(account_id: int, phone: str) -> dict | None:
    """
    Busca um contato pelo número de telefone (wa_id).
    Retorna o primeiro resultado ou None se não encontrado.
    """
    url = f"{_base(account_id)}/contacts/search"
    async with httpx.AsyncClient() as client:
        r = await client.get(url, headers=_headers(), params={"q": phone, "page": 1})
        r.raise_for_status()
        data = r.json()
        results = data.get("payload", [])
        return results[0] if results else None


async def get_contact_conversations(account_id: int, contact_id: int) -> list[dict]:
    """
    Retorna as conversas de um contato, ordenadas por última atividade.
    """
    url = f"{_base(account_id)}/contacts/{contact_id}/conversations"
    async with httpx.AsyncClient() as client:
        r = await client.get(url, headers=_headers())
        r.raise_for_status()
        payload = r.json().get("payload", [])
        # Ordena por last_activity_at desc para pegar a conversa mais recente
        return sorted(payload, key=lambda c: c.get("last_activity_at", 0), reverse=True)


async def send_message_to_phone(phone: str, content: str) -> bool:
    """
    Envia uma mensagem de texto para um número de telefone via a conversa mais recente no Chatwoot.
    Retorna True se enviado com sucesso, False se o contato não foi encontrado.
    """
    account_id = settings.CHATWOOT_ACCOUNT_ID
    contact = await search_contact_by_phone(account_id, phone)
    if not contact:
        return False

    contact_id = contact["id"]
    conversations = await get_contact_conversations(account_id, contact_id)
    if not conversations:
        return False

    conversation_id = conversations[0]["id"]
    await send_message(account_id, conversation_id, content)
    return True

File: src/main/kotlin/ai/koalla/core/client/ChatwootClient.kt
package ai.koalla.core.client

import ai.koalla.core.config.KoallaProperties
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitBodyOrNull
import org.springframework.web.reactive.function.client.awaitExchange

/**
 * Chatwoot API client — mirrors all HTTP calls from the Python version.
 */
@Component
class ChatwootClient(
    private val chatwootWebClient: WebClient,
    private val props: KoallaProperties
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    private fun basePath(accountId: Int = props.chatwoot.accountId): String =
        "/api/v1/accounts/$accountId"

    /**
     * POST update_last_seen — marks messages as read.
     */
    suspend fun markAsRead(accountId: Int, conversationId: Int) {
        try {
            chatwootWebClient.post()
                .uri("${basePath(accountId)}/conversations/$conversationId/update_last_seen")
                .retrieve()
                .awaitBodyOrNull<Unit>()
        } catch (e: Exception) {
            logger.warn("Failed to mark conversation $conversationId as read: ${e.message}")
        }
    }

    /**
     * POST a text message to a conversation.
     */
    suspend fun sendMessage(accountId: Int, conversationId: Int, content: String): Map<String, Any>? {
        return try {
            chatwootWebClient.post()
                .uri("${basePath(accountId)}/conversations/$conversationId/messages")
                .bodyValue(mapOf("content" to content))
                .retrieve()
                .awaitBody<Map<String, Any>>()
        } catch (e: Exception) {
            logger.error("Failed to send message to conversation $conversationId: ${e.message}")
            null
        }
    }

    /**
     * POST a reaction to a specific message.
     */
    suspend fun sendReaction(
        accountId: Int,
        conversationId: Int,
        messageId: Int,
        emoji: String
    ): Map<String, Any>? {
        val body = mapOf(
            "content" to emoji,
            "content_attributes" to mapOf(
                "in_reply_to" to messageId,
                "is_reaction" to true
            )
        )
        return try {
            chatwootWebClient.post()
                .uri("${basePath(accountId)}/conversations/$conversationId/messages")
                .bodyValue(body)
                .retrieve()
                .awaitBody<Map<String, Any>>()
        } catch (e: Exception) {
            logger.error("Failed to send reaction: ${e.message}")
            null
        }
    }

    /**
     * POST/replace conversation labels.
     */
    suspend fun updateLabels(accountId: Int, conversationId: Int, labels: List<String>) {
        try {
            chatwootWebClient.post()
                .uri("${basePath(accountId)}/conversations/$conversationId/labels")
                .bodyValue(mapOf("labels" to labels))
                .retrieve()
                .awaitBodyOrNull<Unit>()
        } catch (e: Exception) {
            logger.warn("Failed to update labels: ${e.message}")
        }
    }

    /**
     * GET contact details (includes custom_attributes).
     */
    suspend fun getContact(accountId: Int, contactId: Int): Map<String, Any>? {
        return try {
            chatwootWebClient.get()
                .uri("${basePath(accountId)}/contacts/$contactId")
                .retrieve()
                .awaitBody<Map<String, Any>>()
        } catch (e: Exception) {
            logger.warn("Failed to get contact $contactId: ${e.message}")
            null
        }
    }

    /**
     * PATCH contact custom attributes.
     */
    suspend fun updateContactAttributes(
        accountId: Int,
        contactId: Int,
        customAttributes: Map<String, Any>
    ): Map<String, Any>? {
        return try {
            chatwootWebClient.patch()
                .uri("${basePath(accountId)}/contacts/$contactId")
                .bodyValue(mapOf("custom_attributes" to customAttributes))
                .retrieve()
                .awaitBody<Map<String, Any>>()
        } catch (e: Exception) {
            logger.error("Failed to update contact attributes: ${e.message}")
            null
        }
    }

    /**
     * POST destroy_custom_attributes.
     */
    suspend fun destroyContactAttributes(
        accountId: Int,
        contactId: Int,
        attributes: List<String>
    ) {
        try {
            chatwootWebClient.post()
                .uri("${basePath(accountId)}/contacts/$contactId/destroy_custom_attributes")
                .bodyValue(mapOf("custom_attributes" to attributes))
                .retrieve()
                .awaitBodyOrNull<Unit>()
        } catch (e: Exception) {
            logger.warn("Failed to destroy contact attributes: ${e.message}")
        }
    }

    /**
     * Download a binary attachment (audio, file) from Chatwoot storage.
     */
    suspend fun downloadAttachment(dataUrl: String): ByteArray? {
        return try {
            chatwootWebClient.get()
                .uri(dataUrl)
                .retrieve()
                .awaitBody<ByteArray>()
        } catch (e: Exception) {
            logger.error("Failed to download attachment: ${e.message}")
            null
        }
    }

    /**
     * Search contact by phone number.
     */
    suspend fun searchContactByPhone(accountId: Int, phone: String): Map<String, Any>? {
        return try {
            val response = chatwootWebClient.get()
                .uri("${basePath(accountId)}/contacts/search?q=$phone&page=1")
                .retrieve()
                .awaitBody<Map<String, Any>>()

            @Suppress("UNCHECKED_CAST")
            val payload = response["payload"] as? List<Map<String, Any>>
            payload?.firstOrNull()
        } catch (e: Exception) {
            logger.warn("Failed to search contact by phone $phone: ${e.message}")
            null
        }
    }

    /**
     * Get contact conversations, sorted by last activity.
     */
    suspend fun getContactConversations(accountId: Int, contactId: Int): List<Map<String, Any>> {
        return try {
            val response = chatwootWebClient.get()
                .uri("${basePath(accountId)}/contacts/$contactId/conversations")
                .retrieve()
                .awaitBody<Map<String, Any>>()

            @Suppress("UNCHECKED_CAST")
            val payload = response["payload"] as? List<Map<String, Any>> ?: emptyList()
            payload.sortedByDescending { it["last_activity_at"] as? Long ?: 0L }
        } catch (e: Exception) {
            logger.warn("Failed to get contact conversations: ${e.message}")
            emptyList()
        }
    }

    /**
     * Send message to a phone number via the most recent conversation.
     */
    suspend fun sendMessageToPhone(phone: String, content: String): Boolean {
        val accountId = props.chatwoot.accountId
        val contact = searchContactByPhone(accountId, phone) ?: return false

        val contactId = (contact["id"] as? Number)?.toInt() ?: return false
        val conversations = getContactConversations(accountId, contactId)
        if (conversations.isEmpty()) return false

        val conversationId = (conversations[0]["id"] as? Number)?.toInt() ?: return false
        sendMessage(accountId, conversationId, content)
        return true
    }
}

